### ISSUE TITLE
Wide Review Comment 2017: position and size reference is video or video viewport?

### ISSUE THREAD

**Issue by nigelmegitt** [w3c/webvtt#375]
Copy/paste from https://lists.w3.org/Archives/Public/public-tt/2017Sep/0080.html - raising as an issue for tracking/disposition purposes.

In some places sizes and positions are defined relative to the video viewport; in others the video itself. This is likely to cause some confusion or mis-alignment when the two are not the same (e.g. a 16:9 aspect ratio video is displayed in a 14:9 viewport) and creates an authoring problem. For example the cue box size is relative to the video but the cue box line is relative to the video viewport (both defined within section 3.1).

**Comment by nigelmegitt** [comment]
Additional point on this - there are some places where the `vw` and `vh` CSS units are used for positioning and sizing, yet they do not seem to accommodate scenarios where video is not shown full-window. Note that `vw` and `vh` are always relative to the initial containing block. 

I'm happy to be shown otherwise but I don't believe that CSS currently has a concept for setting a local viewport within which the `vw` and `vh` units are resolved. This means that the rule:

> The viewport (and initial containing block) is video’s rendering area.

does not actually seem to be implementable - I see that there's an open Chromium bug on this at https://bugs.chromium.org/p/chromium/issues/detail?id=244618 raised by people who found this a lot earlier than me.

**Comment by silviapfeiffer** [comment]
About first comment: Well spotted - "a percentage of the video" means nothing - it should indeed be "video viewport".

About second comment: The way CSS is defined, vh and vw refer to the viewport, which is the "containing block" as per https://www.w3.org/TR/CSS21/visudet.html#containing-block-details . This also states that for media elements the "initial containing block" is the video viewport, so all WebVTT dimensions are calculated with respect to the block provided by the video. That is independent of whether the video is fullscreen or not - it's always relative to the actual rendered dimensions of the video. The sentence you quoted is merely repeating what the CSS spec says.

The bug that you are referring to simply says that Chromium's implementation of vh and vw for all shadow DOM elements of the video element (and that includes WebVTT rendering) is broken.

**Comment by dwsinger** [comment]
are vw and vh established by the containing context (e.g. HTML) if any, or are they reset by the video element? If the latter, then it won't be possible to be consistent in e.g. font sizing. The phrase "The viewport (and initial containing block) is video’s rendering area." essentially sets a context for  this instantiation of CSS.  Perhaps it needs to say "The CSS viewport (as defined in CSS 2.1 section XX)..."

**Comment by dwsinger** [comment]
Do we need to say that the root element from CSS's point of view is the video viewport, to be clear? Or the initial containing block?

**Comment by silviapfeiffer** [comment]
Resolution at TPAC:

This is fundamental to the specification of WebVTT:
"The viewport (and initial containing block) is video’s rendering area."
It needs to be in a better location and explain how `vh` and `vw` are interpreted in this context.

It may even be important to clarify this for CSS in general, because the video shadow DOM should apply only to the video element's rendering area, not the full Web page - should be discussed with CSS WG?

**Comment by dwsinger** [comment]
Note that this means that a naive javascript implementation is hard, because it's not easy to establish a new CSS sandbox/context?

**Comment by dwsinger** [comment]
(We need to loop back around with CSS)

**Comment by silviapfeiffer** [comment]
Did you get to talk to the CSS group about this?

**Comment by fantasai** [comment]
The viewport units in CSS are currently resolved to a length throughout the entire tree, and are computed before layout begins. They thus cannot depend on layout happening in that document. If you've got an iframe, then the referenced document has its own viewport and rendering tree, so it resolves everything independently based on the size of the iframe.

I don't quite understand the relationship of WebVTT to the rest of the document, so I can't really comment further on what should happen.

Wrt https://github.com/w3c/webvtt/issues/375#issuecomment-343597933 there is a very significant distinction between “a containing block” (one per element, usually but not always determined by the dimensions of the parent) and “the initial containing block” (one per document, established by the dimensions of the page viewport, the same one against which media queries are resolved).

**Comment by dwsinger** [comment]
Thank you. I think perhaps the spec is trying to say then that the video region is like an iFrame, with sizes resolving relative to that. Maybe we could express that more clearly?

Dave Singer (iPhone)

**Comment by silviapfeiffer** [comment]
@fantasai thanks for the clarification! That will also have impact on https://bugs.chromium.org/p/chromium/issues/detail?id=244618 .

@dwsinger yes, we need to change the wording from "initial containing block" to "the block defined by the video element" or some such.

**Comment by nigelmegitt** [comment]
Makes sense to me too. Note that in TTML2 we defined `rw` and `rh` units to be percentage of the video width and height respectively. Perhaps we could align and use the same concept here?

**Comment by nigelmegitt** [comment]
Seems like there's more change needed here than just what is in #240 now, to address https://github.com/w3c/webvtt/issues/375#issuecomment-350415818.

**Comment by dwsinger** [comment]
You want more changes of the term "video viewport"?

**Comment by silviapfeiffer** [comment]
Just looking at #379 as well. So maybe I can add to the definition of "video viewport" that in cases of overscan, the "video viewport" is limited to the rendered area of the video? (I note that TTML2 does not mention overscan.)

**Comment by nigelmegitt** [comment]
> Would it be ok to continue using "video viewport" and make a note that in WebVTT, the video element's width and height equate to the "" width and height of the "root container region" as established by the "related media object region" "" in TTML2?

@silviapfeiffer Depends what you mean by ok! If you're describing the current intended behaviour in WebVTT, then yes, keep using "video viewport", and yes, add the comment about overscan, but also note that it goes further than overscan, since the aspect ratio of the video might not be the same as the video element, and the implementation might do some area filling zoom behaviour.

However the proposed note referencing those TTML2 terms (great idea to add that, by the way) is not correct. The reason the two distinct definitions exist in TTML2 is because the related media object region does not need to be coincident with the root container region. I'm not 100% clear how the related media object region corresponds to the video element, since the HTML spec doesn't seem to explain what happens when the `video` element's `width` and `height` attributes do not have the same ration as the video's intrinsic width and height, but it's bound to be the closest concept.

The way this _should_ work in my view is that spatial coordinates should relate to the _video_ and the implementation should make some decisions about how to handle the case where not all the video is shown in the video viewport and therefore some cropping of text might occur. That is the only way to author accessible text, since there has to be some way to position relative to the images in the underlying video, otherwise mouths, in-video text etc that is important may be obscured. Note that IMSC 1.0.1 has introduced the concept of [active area](https://www.w3.org/TR/ttml-imsc1.0.1/#ittp-activeArea) to signal where there is editorially important text displayed, and help the implementation make those decisions in a non-prescriptive way.

**Comment by silviapfeiffer** [comment]
The video in HTML will scale into the width and height of the video element. If the scale is not workable, it will introduce transparent space for letterboxing or pillarboxing. You can choose the color using background-color.

We cannot make the coordinates position on the intrinsic video's dimensions, because the rendering area for the WebVTT captions provided by HTML is the video element's dimensions. It's really up to the Web developer to make sure that there is no letter- or pillarboxing when displaying videos with captions.

**Comment by nigelmegitt** [comment]
> We cannot make the coordinates position on the intrinsic video's dimensions, because the rendering area for the WebVTT captions provided by HTML is the video element's dimensions.

I believe there should be a way to map to the video's intrinsic dimensions, as follows:

According to http://w3c.github.io/html/semantics-embedded-content.html#the-video-element :

> When a video element’s media resource has a video channel, the element provides a paint source whose width is the media resource’s intrinsic width, whose height is the media resource’s intrinsic height, [snip]

and

> In the absence of style rules to the contrary, video content should be rendered inside the element’s playback area such that the video content is shown centered in the playback area at the largest possible size that fits completely within it, with the video content’s aspect ratio being preserved. Thus, if the aspect ratio of the playback area does not match the aspect ratio of the video, the video will be shown letterboxed or pillarboxed. Areas of the element’s playback area that do not contain the video represent nothing.

So it seems that there is at least an intent of allowing the video element to be aware of, and display, the video content's intrinsic dimensions. Further, the [HTMLVideoElement](http://w3c.github.io/html/semantics-embedded-content.html#htmlvideoelement) provides accessors to get the intrinsic size as well as the video window size.

**Comment by silviapfeiffer** [comment]
It's not how any of the browsers have implemented the caption rendering. And why would they? If your captions can be rendered over a black letterboxed area rather than on top of the video, wouldn't that be preferable to the users? It won't obstruct your viewing area. If the web page author has decided to dimension the video on this way, why should the browser insist on drawing on top of the video? Caption without intent is provided with relation to the intrinsic video dimensions but overruled by the web page author and that's perfectly ok.

**Comment by tairt** [comment]
As always (also in TTML) the most basic questions are the most difficult. 

@silviapfeiffer Could you make a PR based on what the intent was and how it has been interpreted by the web browser.

**Comment by silviapfeiffer** [comment]
I've analysed how CSS deals with the object size of replaced content. It's all described with a lot of detail here: https://www.w3.org/TR/css3-images/#sizing . Thus, what we have called "video viewport" directly relates to the "concrete object size" of the HTML5 video element. I've thus added a definition of "video viewport" that makes this explicit. That's in the pull request #434.

We could go one step further and add an attribute to WebVTT or to the HTML5 track element that allows an alternative definition of the "video viewport" which would require the browser to render the captions & subtitles onto the video's "intrinsic dimensions" rather than the "concrete object size". I believe this should be done together with the HTML5 WG and may quite possibly require that the term "video viewport" to be moved into the HTML5 spec. It will definitely require discussions with browser vendors. I'd recommend postponing this to WebVTTv2.

**Comment by nigelmegitt** [comment]
@silviapfeiffer having clarified that video viewport aligns to the concrete object size of the video element, which makes sense, how do we deal with the incorrect use of `vw` and `vh` units? One option would be to say something like "for the purposes of resolving `vh` and `vw` units an element whose initial containing block has the same size and dimensions as the video viewport is required, for example a suitably positioned `iframe`". That could be in the new section 3.2 Video Viewport as introduced in #434. What do you think?

**Comment by silviapfeiffer** [comment]
@nigelmegitt good catch - I had totally forgotten about `vh` and `vw`. I think we need to rename them completely so as not to conflict with CSS.
Just re-reading fantasai's comment, we also have to rename "initial containing block".

I'll prepare a fix.

**Comment by nigelmegitt** [comment]
Thanks @silviapfeiffer - can I suggest aligning with [TTML2](https://www.w3.org/TR/2018/CR-ttml2-20180313/#style-value-length) here, and using `rh` and `rw`?

**Comment by silviapfeiffer** [comment]
I was going to align with TTML2, but I don't think the definition of "root container region" and "video viewport" match exactly, particularly since a "root container region" has no border, padding etc. I don't think "root container region" maps easily into CSS and HTML's view of the world, so I didn't want to confuse it. I'd be happy to take advice on this though.

**Comment by nigelmegitt** [comment]
The way I see it the "root container region" is the basis for the spatial coordinate system, and corresponds externally to some real world rectangle at presentation time. `0% 0%` is the top left and `100% 100%` is the bottom right. Padding and border aren't available on it, but may have been taken into account when positioning that real world rectangle - padding and border are available on regions though. 

Compare with WebVTT, where there is an identical internal coordinate system, having `(0, 0)` as the top left and `(100, 100)` as the bottom right, and all positions are relative to it. There are two additional constraints in WebVTT relative to TTML:

1. an external constraint, is that the coordinate system is spatially aligned with the concrete object size of the video element. This is akin to the IMSC constraint that positions the root container region relative to the _related video object_, though more prescriptive about what that video object is.

2. an internal constraint, (a syntactic one?) that no coordinates can spill outside the 0-100 range in either dimension, so all text must be positioned inside it. This is akin to the IMSC constraint that [all regions shall not extend beyond the root container region](https://www.w3.org/TR/ttml-imsc1.0.1/#region-dim-pos).

I find it helpful to separate the coordinate system from the positioning of that system against some other object, so I would be in favour of defining `rw` and `rh` as relative units, and then defining the basis for those relative units separately - plus that gives freedom later to allow the use of other elements than `video` if that should be needed, without having the coordinate system nominally bound to the HTML element.

It looks like [CSS Units Level 3](https://www.w3.org/TR/css-values/) does not define any (non initial) containing block relative lengths. I do not know if there are technical reasons for that or if there has just never been a requirement; we should take it to CSS WG and ask them.

**Comment by silviapfeiffer** [comment]
I'm not sure I fully understand. I think what you're trying to say is that there is a container with a coordinate system for positioning things inside it (which is the video viewport). But then you also have another coordinate system and container that contains that first container?

**Comment by nigelmegitt** [comment]
Not quite - I'm saying there are:
* a container `R` with a coordinate system for positioning things inside it, against which content must be authored, which is the Root Container Region and 
* the video viewport `VV` and
*the intrinsic size of the video `VI`,

which are all in principle independent, but from an authoring perspective `R` is usually coincident with `VI` and from a presentation perspective the current assumption in WebVTT is that `R` is coincident with `VV`. 

As an aside to this discussion, if, as sometimes happens, `VV` is not coincident with `VI` then there's a discrepancy to resolve when working out how to map `R` to `VV`: the impact of this mismatch might be that `R` no longer matches the `VI` used for authoring and that positions may be unexpected. 

Back from my aside, separating `R` from `VV` definitionally is helpful since authoring and presentation practices may not match, since the player size is not under the subtitle author's control.

**Comment by silviapfeiffer** [comment]
OK, that use of `R` and `VI` and `VV` isn't going to work for WebVTT, because the HTML spec specifies the intrinsic size of the video and maps it into the container, which is the video viewport. That container is all we can deal with in the WebVTT spec, because we don't know how the Web page author has positioned the video in the container.

For that reason, we actually have to assume `R` is `VV`.

**Comment by silviapfeiffer** [comment]
Could I get a review of #434 so we can land it?

**Comment by nigelmegitt** [comment]
> isn't going to work for WebVTT, because the HTML spec specifies the intrinsic size of the video and maps it into the container, which is the video viewport.

I'm not sure how easy this is to resolve - I take your point that it might need to be looked at in the context of the HTML spec. Unfortunately "this isn't going to work for WebVTT" is a weaker reason than "this isn't going to work for the user", and in this case, the result is not going to work for the user reliably. That's a problem.

**Comment by silviapfeiffer** [comment]
On the Web, the person that has to make sure it works for the user is and has always been the Web developer. All the Web platform can provide is the tools to make it work. WebVTT is no different and there's nothing here that stops a Web developer from making it work for a user.

Given this, #434 should land as it explains how the video viewport works.

**Comment by nigelmegitt** [comment]
I'm reviewing #434. 

> On the Web, the person that has to make sure it works for the user is and has always been the Web developer.

This is not at all true for any content generated externally and inserted into a page, including UGC (text comments on blog pages, say) and videos and subtitles/captions uploaded to generic hosting sites. In the latter case the uploader may want to post the same content to different sites, and should not need to adjust it based on the whims of the sites' Web developers, in the unlikely case that those whims were even available.

**Comment by Den4ik98** [comment]
![f2ZxkbjzI2Q](https://user-images.githubusercontent.com/88166702/236616548-e85502c5-5291-4c69-9a08-52e118a0e66b.jpg)
